### ISSUE TITLE
Code Review

### ISSUE THREAD

**Issue by Stephen-Seo** [riggskevinp/InvertedPendulum#1]
Honestly I'm not that familiar programming with the QT framework, so it was difficult for me to grasp the purpose of the objects you've defined in your projects. Though I could tell there was some kind of "tree hierarchy" relationship between some objects (like between the "cart" and the "pendulum").

# Dragging?

Looking at the following snippet:

https://github.com/riggskevinp/InvertedPendulum/blob/2d5fcd1775ae642014e8b8d4b4ef04f98f564b6b/robot.h#L81-L97

The `dragEnterEvent(...)` and related functions made me think one could interact with the pendulum by clicking and dragging on it, but that didn't seem to work for me. I guess it's another QT related member function(s) that were necessary for the entire structure to function.

(AFTER SOME TESTING): Ok, so I realize now that the Inverted Pendulum wouldn't support interactivity since 1000 steps are calculated for the cart to move to keep the pendulum level.

# Eigen

So it also looks like a lot of the heavy lifting is done by the "Eigen" library which seems to be very useful for your use case. It looks like Eigen uses CUDA to accelerate the simulation (handling matrices for you I think) which is pretty nice.

# Comments?

Since the project is still in an intermediate state (not quite finished yet I presume), there is a lack of comments in the code explaining the purposes and the whats/whys. Hopefully the final result will have populated the code with useful comments, as currently the lack of comments makes it a bit hard to follow what's going on (especially for someone who isn't familiar with QT).

# Parameters

It is neat to tweak the parameters to see the subtle changes in the simulation, but if I am understanding this correctly, the lines printed out by the program print the state of the simulation per-millisecond (though the playback of the simulation appears to be different than "simulated-time") and the display plays back the inverted pendulum's state as printed. If I am understanding this correctly, this means that the simulation "ends" after N states (yes it appears to crash upon 1000 steps (seconds?)). Well, if my understanding of `executeSimulation()` and `SimulateTrajectory(...)` in `InvertedPendulumSim.cpp` is correct.

(After some looking around): Ok, so the following line:

https://github.com/riggskevinp/InvertedPendulum/blob/2d5fcd1775ae642014e8b8d4b4ef04f98f564b6b/InvertedPendulumSim.cpp#L87

Causes the vector of `TrajectoryTuple`s to skip 99 out of every 100 trajectory items. This reduces the size N (100000) vector of `TimeAndState` tuples to 1000 steps. I think that this speeds up the simulation of the inverted pendulum as otherwise the simulation would be 100 times slower (as there are 100 times more steps to simulate). This doesn't mess up the simulation because everything appears to be pre-calculated before displaying via QT. So in other words, the entire simulation is calculated before-hand, then displayed using QT (if I'm understanding this correctly).

# End Notes

The concept of the LQR is hard for me to understand, but you've used it to simulate a physical system (Inverted Pendulum) that works with the parameters given to it (like mass, friction, etc.). More graceful handling of the end state would be nice as it currently just crashes due to `trajectory.at(index)` being out of range of the 1000 steps that the cart/pendulum moves to show the maintaining of the pendulum's upright state. The ability to tweak the params is nice, and it was an interesting choice to set the params via file instead of program arguments.

Looks pretty good, just needs maybe some cleanup and comments.

**Comment by riggskevinp** [comment]
Great comments, I truly appreciate the time you took to give this feedback.

Dragging and interacting would be an awesome 2.0 feature.  You are right it is not a dynamic simulation, it runs once and the visualization demos the trajectory.  I'd really like to do that next though.

Eigen does use hardware acceleration but I don't think it is for how I'm using it.  My hunch is because the matrices/vectors are so small it doesn't use the available hardware to do the linear algebra.  To account for this I'm forcing in some Cuda usage that doesn't quite fit.  My goal was to actually calculate K using cuda, which is just done once, which might still be faster with Eigen but at least I'd have the guarantee that Cuda is used.

Based on your feedback, I'm going to go through and try to use comments judiciously.  I like to try to have code that is very self describing and simple but if it is not immediately clear then I need to clean it.

So the other way to control the simulation playback speed is how long the timer runs for, I think in milliseconds.  A future change would pull the system parameters into the main application and have the timer value set there.

The end state is something I'm still not clear on the path forward.  It would be easy to have it not crash but I haven't touched it because I want to figure out how many simulations I want to run and if I want the simulation to be interactive.  Things like that.
